shopinvader_customer_multi_user/services/user_management.py

```diff
@@ -0,0 +1,201 @@
+
+    def _post_create(self, record):
+        pass
+
+    def _post_update(self, record):
+        pass
+
+    def _validator_search(self):
+        return {
+            "id": {"coerce": to_int, "type": "integer"},
+            "per_page": {
+                "coerce": to_int,
+                "nullable": True,
+                "type": "integer",
+            },
+            "page": {"coerce": to_int, "nullable": True, "type": "integer"},
+            "scope": {"type": "dict", "nullable": True},
+        }
+
+    def _validator_create(self):
+        return {
+            "name": {"type": "string", "required": True},
+            "email": {"type": "string", "required": True},
+            "can_manage_users": {
+                "type": "boolean",
+                "coerce": to_bool,
+                "nullable": True,
+            },
+            "parent_id": {
+                "type": "integer",
+                "coerce": to_int,
+                "nullable": True,
+            },
+        }
+
+    def _validator_update(self):
+        res = self._validator_create()
+        # Prevent email/login update
+        res.pop("email", None)
+        for key in res:
+            if "required" in res[key]:
+                del res[key]["required"]
+        res["state"] = {
+            "type": "string",
+            "required": False,
+            "allowed": ALL_STATES,
+        }
+        return res
+
+    def _get_base_search_domain(self):
+        if (
+            not self._is_logged_in()
+            or not self.invader_partner_user.is_users_manager
+        ):
+            return expression.FALSE_DOMAIN
+        # Simple users w/ delegated permissions
+        # can see only records they own.
+        delegated = (
+            self.invader_partner_user.is_invader_user
+            and self.invader_partner_user.can_manage_users
+        )
+        return self.invader_partner_user._make_partner_domain(
+            "main_partner_id" if not delegated else "parent_id"
+        )
+
+    def _prepare_params(self, params, mode="create"):
+        # Always enforce the backend
+        params["backend_id"] = self.shopinvader_backend.id
+        parent_id = params.pop("parent_id", None)
+        if parent_id:
+            # Users can be created only in the same hierarchy
+            record = (
+                self.env["res.partner"]
+                .browse(parent_id)
+                .filtered_domain(
+                    [
+                        (
+                            "id",
+                            "child_of",
+                            self.invader_partner_user.main_partner_id.id,
+                        )
+                    ]
+                )
+                .exists()
+            )
+            if not record:
+                parent_id = None
+        params["parent_id"] = parent_id or self.partner_user.id
+        return params
+
+    def _to_json(self, records, **kw):
+        return records.jsonify(self._json_parser(), **kw)
```

Comment:
There is something wrong with keywords found here, as jsonify is not intend to find any kw.
```2021-05-05 13:40:29,924 1 ERROR devel odoo.addons.base_rest.http: RESTFULL call to url http://odoo:8069/shopinvader/users?page=1&per_page=10 with method GET and params {'page': '1', 'per_page': '10'} raise the following error 500 Internal Server Error: jsonify() got an unexpected keyword argument 'page' 
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/base_rest/http.py", line 184, in _handle_exception
    return super(HttpRestRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 750, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 310, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/opt/odoo/custom/src/odoo/odoo/tools/pycompat.py", line 14, in reraise
    raise value
  File "/opt/odoo/custom/src/odoo/odoo/addons/base/models/ir_http.py", line 234, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 809, in dispatch
    r = self._call_function(**self.params)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 350, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 915, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/custom/src/odoo/odoo/http.py", line 515, in response_wrap
    response = f(*args, **kw)
  File "<string>", line 6, in get_search
  File "/opt/odoo/auto/addons/base_rest/controllers/main.py", line 144, in _process_method
    result = service.dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 247, in dispatch
    res = super().dispatch(method_name, *args, params=params)
  File "/opt/odoo/auto/addons/base_rest/components/service.py", line 158, in dispatch
    res = method(*args, **secure_params)
  File "/opt/odoo/auto/addons/base_rest/restapi.py", line 61, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 39, in wrapped
    return func(*args, **kwargs)
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 64, in search
    return self._paginate_search(**params)
  File "/opt/odoo/auto/addons/shopinvader/services/service.py", line 141, in _paginate_search
    return {"size": total_count, "data": self._to_json(records, **params)}
  File "/opt/odoo/auto/addons/shopinvader_customer_multi_user/services/user_management.py", line 181, in _to_json
    return records.jsonify(self._json_parser(), **kw)
TypeError: jsonify() got an unexpected keyword argument 'page'
```

Reply:
Error related with this merge: https://github.com/shopinvader/odoo-shopinvader/pull/942 @simahawk

Reply:
jsonify accepts kw but not random ones. It happens only when extra params are passed (that's why tests are green). There's something wrong w/ a change I recently made to propagate params across calls. Here we definitely need to sanitize them. I'll check, thanks!

Reply:
> Error related with this merge: #942 @simahawk

yes, exactly, I just have to fix it here for user_management

Reply:
last commit should fix it

Reply:
@HviorForgeFlow can you confirm it's working now?